File: packages/asbr/src/connectors/manifest.ts
import { readFile } from 'node:fs/promises';
import { resolve } from 'node:path';
import { fileURLToPath } from 'node:url';
import {
	type ConnectorManifestEntry,
	type ConnectorServiceEntry,
	ConnectorServiceEntrySchema,
	type ConnectorServiceMap,
	type ConnectorServiceMapPayload,
	ConnectorServiceMapPayloadSchema,
	ConnectorServiceMapSchema,
	type ConnectorsManifest,
	ConnectorsManifestSchema,
} from '@cortex-os/asbr-schemas';
import { assertManifestDocument } from './schema.js';
import { signConnectorPayload } from './signature.js';

const BRAND = 'brAInwav' as const;
const MODULE_DIR = fileURLToPath(new URL('.', import.meta.url));
const FALLBACK_MANIFEST_PATH = resolve(MODULE_DIR, '../../../../config/connectors.manifest.json');
const WORKING_DIR_MANIFEST_PATH = resolve(process.cwd(), 'config', 'connectors.manifest.json');

export interface ManifestLoadAttempt {
	path: string;
	error: unknown;
}

export class ConnectorsManifestError extends Error {
	constructor(
		message: string,
		public readonly attempts: ManifestLoadAttempt[],
	) {
		super(message);
		this.name = 'ConnectorsManifestError';
	}
}

interface BuildConnectorServiceMapOptions {
	now?: () => Date;
}

export async function loadConnectorsManifest(manifestPath?: string): Promise<ConnectorsManifest> {
	const attempts: ManifestLoadAttempt[] = [];

	for (const candidate of buildCandidatePaths(
		manifestPath ?? process.env.CONNECTORS_MANIFEST_PATH,
	)) {
		try {
			const raw = await readFile(candidate, 'utf-8');
			const parsed = JSON.parse(raw) as unknown;
			assertManifestDocument(parsed);
			return ConnectorsManifestSchema.parse(parsed);
		} catch (error) {
			attempts.push({ path: candidate, error });
		}
	}

	throw new ConnectorsManifestError(
		'Unable to load connectors manifest from configured locations',
		attempts,
	);
}

export function buildConnectorServiceMap(
	manifest: ConnectorsManifest,
	options: BuildConnectorServiceMapOptions = {},
): ConnectorServiceMapPayload {
	const now = options.now?.() ?? new Date();
	const generatedAt = manifest.generatedAt ?? now.toISOString();

	const connectors = manifest.connectors
		.map((connector) => buildConnectorEntry(connector))
		.sort((left, right) => left.id.localeCompare(right.id));

	if (connectors.length === 0) {
		throw new ConnectorsManifestError(
			'Connectors manifest must include at least one connector',
			[],
		);
	}

	const connectorTtls = connectors.map((entry) => entry.ttlSeconds);
	const minConnectorTtl = connectorTtls.length ? Math.min(...connectorTtls) : 1;
	const manifestTtl = manifest.ttlSeconds ?? minConnectorTtl;
	const ttlSeconds = Math.max(1, Math.min(manifestTtl, minConnectorTtl));

	const payload = {
		id: manifest.id,
		brand: manifest.brand ?? BRAND,
		generatedAt,
		ttlSeconds,
		connectors,
		...(manifest.metadata ? { metadata: { ...manifest.metadata } } : {}),
	};

	return ConnectorServiceMapPayloadSchema.parse(payload);
}

export function signConnectorServiceMap(
	payload: ConnectorServiceMapPayload,
	secret: string,
): string {
	if (!secret) {
		throw new Error('CONNECTORS_SIGNATURE_KEY is required to sign the connectors service map');
	}

	return signConnectorPayload(payload, secret);
}

export function attachSignature(
	payload: ConnectorServiceMapPayload,
	signature: string,
): ConnectorServiceMap {
	return ConnectorServiceMapSchema.parse({ ...payload, signature });
}

function buildConnectorEntry(connector: ConnectorManifestEntry): ConnectorServiceEntry {
	const enabled = resolveEnabled(connector);
	const auth = connector.auth ?? deriveAuthFromAuthentication(connector);
	const headers = mergeHeaders(connector);
	const quotas = normalizeQuotas(connector);
	const timeouts = connector.timeouts ? { ...connector.timeouts } : undefined;
	const metadata = {
		brand: BRAND,
		...(connector.metadata ?? {}),
	};

	// Deep clone remoteTools if present and non-empty
	const remoteTools =
		connector.remoteTools && connector.remoteTools.length > 0
			? connector.remoteTools.map((tool) => ({
					...tool,
					...(tool.tags ? { tags: [...tool.tags] } : {}),
					...(tool.scopes ? { scopes: [...tool.scopes] } : {}),
				}))
			: undefined;

	const baseEntry: ConnectorServiceEntry = {
		id: connector.id,
		version: connector.version,
		displayName: connector.displayName ?? connector.name,
		endpoint: connector.endpoint,
		auth,
		scopes: [...connector.scopes],
		ttlSeconds: connector.ttlSeconds,
		enabled,
		metadata,
		...(quotas ? { quotas } : {}),
		...(headers ? { headers } : {}),
		...(connector.description ? { description: connector.description } : {}),
		...(connector.tags && connector.tags.length > 0 ? { tags: [...connector.tags] } : {}),
		...(timeouts ? { timeouts } : {}),
		...(remoteTools ? { remoteTools } : {}),
	};

	return ConnectorServiceEntrySchema.parse(baseEntry);
}

function resolveEnabled(connector: ConnectorManifestEntry): boolean {
	if (typeof connector.enabled === 'boolean') {
		return connector.enabled;
	}

	if (connector.status) {
		return connector.status !== 'disabled';
	}

	return true;
}

function deriveAuthFromAuthentication(
	connector: ConnectorManifestEntry,
): ConnectorServiceEntry['auth'] {
	const authConfig = connector.authentication;
	if (!authConfig || authConfig.headers.length === 0) {
		return { type: 'none' };
	}

	const primaryHeader = authConfig.headers[0];
	const headerName = primaryHeader.name;
	const normalized = headerName.toLowerCase();

	if (normalized === 'authorization' || primaryHeader.value.startsWith('Bearer ')) {
		return { type: 'bearer', headerName };
	}

	return { type: 'apiKey', headerName };
}

function mergeHeaders(
	connector: ConnectorManifestEntry,
): ConnectorServiceEntry['headers'] | undefined {
	const merged = new Map<string, string>();

	const authHeaders = connector.authentication?.headers ?? [];
	for (const header of authHeaders) {
		merged.set(header.name, header.value);
	}

	if (connector.headers) {
		for (const [key, value] of Object.entries(connector.headers)) {
			merged.set(key, value);
		}
	}

	return merged.size > 0 ? Object.fromEntries(merged.entries()) : undefined;
}

function normalizeQuotas(
	connector: ConnectorManifestEntry,
): ConnectorServiceEntry['quotas'] | undefined {
	const quotas = connector.quotas;
	if (!quotas) {
		return undefined;
	}

	const budget: NonNullable<ConnectorServiceEntry['quotas']> = {};

	if (typeof quotas.perMinute === 'number') {
		budget.perMinute = quotas.perMinute;
	} else if (typeof (quotas as Record<string, number>).per_minute === 'number') {
		budget.perMinute = (quotas as Record<string, number>).per_minute;
	}

	if (typeof quotas.perHour === 'number') {
		budget.perHour = quotas.perHour;
	} else if (typeof (quotas as Record<string, number>).per_hour === 'number') {
		budget.perHour = (quotas as Record<string, number>).per_hour;
	}

	if (typeof quotas.perDay === 'number') {
		budget.perDay = quotas.perDay;
	} else if (typeof (quotas as Record<string, number>).per_day === 'number') {
		budget.perDay = (quotas as Record<string, number>).per_day;
	}

	if (typeof quotas.concurrent === 'number') {
		budget.concurrent = quotas.concurrent;
	}

	return Object.keys(budget).length > 0 ? budget : undefined;
}

function buildCandidatePaths(manifestPath?: string): string[] {
	if (manifestPath) {
		return [resolve(manifestPath)];
	}

	const candidates = [WORKING_DIR_MANIFEST_PATH, FALLBACK_MANIFEST_PATH].map((path) =>
		resolve(path),
	);

	return Array.from(new Set(candidates));
}

File: packages/connectors/src/cortex_connectors/manifest.py
"""Utilities for loading and signing the connectors manifest."""

from __future__ import annotations

import json
import hmac
from dataclasses import dataclass
from datetime import datetime, timezone
from hashlib import sha256
from pathlib import Path
from typing import Iterable, List, Optional

from pydantic import ValidationError

from .models import ConnectorServiceMapPayload, ConnectorsManifest

MODULE_PATH = Path(__file__).resolve()
FALLBACK_MANIFEST_PATH = MODULE_PATH.parents[4] / "config" / "connectors.manifest.json"
CWD_MANIFEST_PATH = Path.cwd() / "config" / "connectors.manifest.json"


@dataclass(slots=True)
class ManifestLoadAttempt:
    """Record of an attempted manifest load."""

    path: Path
    error: Exception


class ConnectorsManifestError(Exception):
    """Raised when the manifest cannot be loaded from any configured location."""

    def __init__(self, message: str, attempts: List[ManifestLoadAttempt]):
        super().__init__(message)
        self.attempts = attempts


def _build_candidate_paths(path: Optional[str | Path]) -> Iterable[Path]:
    """Build the ordered list of manifest paths to try."""

    candidates: List[Path] = []
    if path:
        candidates.append(Path(path))
    candidates.extend([CWD_MANIFEST_PATH, FALLBACK_MANIFEST_PATH])

    seen: set[Path] = set()
    for candidate in candidates:
        expanded = candidate.expanduser()
        if expanded in seen:
            continue
        seen.add(expanded)
        yield expanded


def load_connectors_manifest(manifest_path: Optional[str | Path] = None) -> ConnectorsManifest:
    """Load and validate the connectors manifest."""

    attempts: List[ManifestLoadAttempt] = []

    for candidate in _build_candidate_paths(manifest_path):
        try:
            raw = candidate.read_text(encoding="utf-8")
            data = json.loads(raw)
            return ConnectorsManifest.model_validate(data)
        except (FileNotFoundError, PermissionError, json.JSONDecodeError, ValidationError) as exc:  # pragma: no cover - aggregated below
            attempts.append(ManifestLoadAttempt(path=candidate, error=exc))
        except Exception as exc:  # pragma: no cover - unexpected exceptions are still reported
            attempts.append(ManifestLoadAttempt(path=candidate, error=exc))

    raise ConnectorsManifestError(
        "Unable to load connectors manifest from configured locations",
        attempts,
    )


def build_connector_service_map(manifest: ConnectorsManifest) -> ConnectorServiceMapPayload:
    """Create the ASBR-facing service map from the manifest."""

    generated_at = (
        manifest.generated_at
        if manifest.generated_at is not None
        else datetime.now(timezone.utc).isoformat().replace("+00:00", "Z")
    )

    connectors: List[ConnectorServiceMapEntry] = []
    for entry in sorted(manifest.connectors, key=lambda item: item.id):
        is_enabled = entry.enabled
        if entry.status is not None:
            is_enabled = entry.status != "disabled" and entry.enabled

        connector = ConnectorServiceMapEntry(
            id=entry.id,
            version=entry.version,
            display_name=entry.display_name,
            endpoint=entry.endpoint,
            auth=entry.auth,
            scopes=list(entry.scopes),
            ttl_seconds=entry.ttl_seconds,
            enabled=is_enabled,
            metadata=dict(entry.metadata),
            description=entry.description,
            tags=list(entry.tags) if entry.tags else None,
        )

        if entry.quotas:
            connector.quotas = dict(entry.quotas)

        if entry.timeouts:
            connector.timeouts = dict(entry.timeouts)

        connectors.append(connector)

    min_connector_ttl = min((connector.ttl_seconds for connector in connectors), default=1)
    ttl_seconds = max(manifest.ttl_seconds or min_connector_ttl, 1)

    return ConnectorServiceMap(
        id=manifest.id,
        generated_at=generated_at,
        ttl_seconds=ttl_seconds,
        connectors=connectors,
    )
    return manifest.service_map_payload()


def sign_connector_service_map(service_map: ConnectorServiceMapPayload, secret: str) -> str:
    """Generate an HMAC signature for the service map."""

    if not secret:
        msg = "CONNECTORS_SIGNATURE_KEY is required to sign the connectors service map"
        raise ValueError(msg)

    payload = json.dumps(
        service_map.model_dump(mode="json", by_alias=True, exclude_none=True),
        separators=(",", ":"),
        sort_keys=True,
    ).encode("utf-8")
    return hmac.new(secret.encode("utf-8"), payload, sha256).hexdigest()
